### ISSUE TITLE
Modelo de predicción de demanda de préstamos (Gradient Boosting)

### ISSUE THREAD

**Issue by Sebastian906** [Sebastian906/Zibooka-FullStack#180]
| Campo | Valor |
|-------|-------|
| **Tipo** | `feature` |
| **Prioridad** | 🟡 High |
| **Estimación** | 3 días |
| **Depende de** | #179  |
| **Bloquea** | Ninguno |

## Descripción

Implementar un modelo de Gradient Boosting (XGBoost) que prediga qué libros tendrán alta demanda de préstamos en los próximos 30 días. Esto permite al administrador anticipar necesidades de inventario, especialmente relevante en contexto educativo donde la demanda es cíclica (inicios de semestre, épocas de exámenes).

## Objetivo

Tener un endpoint `/predict/demand` que retorne libros con alta probabilidad de demanda en los próximos 30 días, y otro `/train/demand` para entrenar/actualizar el modelo periódicamente.

## Features del Modelo

| Feature | Tipo | Descripción | Fuente en MongoDB |
|---------|------|-------------|-------------------|
| `month` | int (1-12) | Mes del año | `loan.loanDate` |
| `week_of_year` | int (1-52) | Semana del año | `loan.loanDate` |
| `day_of_week` | int (0-6) | Día de la semana (0=domingo) | `loan.loanDate` |
| `is_semester_start` | bin (0/1) | Semanas 1-3 o 16-18 del año | Derivado de loanDate |
| `category_encoded` | int | Categoría del libro (label encoded) | `product.category` |
| `loan_count_prev_30d` | int | Préstamos en últimos 30 días previos a la fecha | Aggregation sobre Loan |
| `loan_count_prev_90d` | int | Préstamos en últimos 90 días | Aggregation sobre Loan |
| `total_loans_all_time` | int | Préstamos totales históricos del título | Aggregation sobre Loan |
| `avg_loan_duration_days` | float | Duración promedio del préstamo en días | Aggregation sobre Loan |
| `days_since_last_loan` | int | Días desde el último préstamo registrado | Derivado |
| `unique_users_loaned` | int | Usuarios distintos que han tomado el libro | Aggregation sobre Loan |
| `stock_available` | bin (0/1) | Si el libro tiene stock disponible | `product.inStock` |

## Target

Variable binaria: **1** si el libro tuvo ≥ `N` préstamos en los 30 días posteriores a la fecha de observación (N configurable, default 3).

### 1. Clase DemandPredictor

**Archivo:** `zibooka-ml/app/models/demand_predictor.py`

```python
import pandas as pd
import numpy as np
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
import joblib
from app.models.base import BasePredictor
from app.database import mongodb

class DemandPredictor(BasePredictor):
    def __init__(self):
        self.model = XGBClassifier(
            n_estimators=200,
            max_depth=6,
            learning_rate=0.1,
            subsample=0.8,
            colsample_bytree=0.8,
            random_state=42,
            use_label_encoder=False,
            eval_metric='logloss'
        )
        self.feature_columns = None
        self.label_encoder = {}  # Para categorías
    
    async def _build_training_data(self) -> pd.DataFrame:
        """Construye dataset de entrenamiento desde MongoDB."""
        # Pipeline de agregación para obtener features por libro por mes
        pipeline = [
            # Unir loans con products
            {
                '$lookup': {
                    'from': 'products',
                    'localField': 'productId',
                    'foreignField': '_id',
                    'as': 'product'
                }
            },
            { '$unwind': '$product' },
            # Solo préstamos devueltos (historial completo)
            { '$match': { 'status': { '$in': ['returned', 'overdue'] } } },
            # Proyectar campos necesarios
            {
                '$project': {
                    'productId': 1,
                    'category': '$product.category',
                    'loanDate': 1,
                    'returnDate': 1,
                    'inStock': '$product.inStock'
                }
            }
        ]
        
        loans_cursor = mongodb.get_collection('loans').aggregate(pipeline)
        loans = await loans_cursor.to_list(length=None)
        
        df = pd.DataFrame(loans)
        # Ingeniería de features (pandas)
        # ... transformaciones temporales y agregaciones
        
        return df
    
    def _engineer_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """Genera features a partir de datos crudos."""
        features = df.copy()
        
        # Features temporales
        features['month'] = pd.to_datetime(features['loanDate']).dt.month
        features['week_of_year'] = pd.to_datetime(features['loanDate']).dt.isocalendar().week
        features['day_of_week'] = pd.to_datetime(features['loanDate']).dt.dayofweek
        
        # Semestre académico
        features['is_semester_start'] = features['week_of_year'].isin([1,2,3,16,17,18]).astype(int)
        
        # Agregaciones por libro
        # ... loan_count_prev_30d, loan_count_prev_90d, etc.
        
        # Label encoding para categorías
        categories = features['category'].unique()
        self.label_encoder = {cat: i for i, cat in enumerate(categories)}
        features['category_encoded'] = features['category'].map(self.label_encoder)
        
        return features
    
    def train(self, training_data: pd.DataFrame) -> dict:
        features = self._engineer_features(training_data)
        
        # Definir target: alta demanda = >= 3 préstamos en 30 días
        # (requiere ventana temporal deslizante)
        X = features[self.feature_columns]
        y = features['high_demand']
        
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42, stratify=y
        )
        
        self.model.fit(
            X_train, y_train,
            eval_set=[(X_test, y_test)],
            verbose=False
        )
        
        # Evaluación
        y_pred = self.model.predict(X_test)
        y_prob = self.model.predict_proba(X_test)[:, 1]
        
        metrics = {
            'accuracy': float(accuracy_score(y_test, y_pred)),
            'precision': float(precision_score(y_test, y_pred)),
            'recall': float(recall_score(y_test, y_pred)),
            'f1_score': float(f1_score(y_test, y_pred)),
            'auc_roc': float(roc_auc_score(y_test, y_prob))
        }
        
        return metrics
    
    def predict(self, features: np.ndarray) -> np.ndarray:
        return self.model.predict_proba(features)[:, 1]  # Probabilidad clase 1
    
    def predict_for_all_books(self) -> list:
        """Predice demanda para todos los libros activos."""
        # Obtener datos actuales de cada libro
        # Construir feature vector para cada uno
        # Retornar lista con scores
        pass
    
    def save(self, path: str) -> None:
        joblib.dump({
            'model': self.model,
            'feature_columns': self.feature_columns,
            'label_encoder': self.label_encoder
        }, path)
    
    @classmethod
    def load(cls, path: str) -> 'DemandPredictor':
        data = joblib.load(path)
        predictor = cls()
        predictor.model = data['model']
        predictor.feature_columns = data['feature_columns']
        predictor.label_encoder = data['label_encoder']
        return predictor
    
    def get_feature_importance(self) -> dict:
        importance = self.model.feature_importances_
        return dict(zip(self.feature_columns, importance.tolist()))
```

### 2. Router de entrenamiento

**Archivo:** `zibooka-ml/app/routers/train.py`

```python
from fastapi import APIRouter, HTTPException
from app.models.demand_predictor import DemandPredictor
from app.config import settings
import os

router = APIRouter()
predictor = DemandPredictor()

@router.post("/demand")
async def train_demand_model():
    """Entrena el modelo de predicción de demanda."""
    try:
        data = await predictor._build_training_data()
        if len(data) < 100:
            raise HTTPException(400, "Datos insuficientes para entrenar (mínimo 100 registros)")
        
        metrics = predictor.train(data)
        
        # Guardar modelo
        model_path = os.path.join(settings.models_dir, "demand_predictor.joblib")
        os.makedirs(settings.models_dir, exist_ok=True)
        predictor.save(model_path)
        
        return {
            "status": "success",
            "samples": len(data),
            "metrics": metrics,
            "feature_importance": predictor.get_feature_importance()
        }
    except Exception as e:
        raise HTTPException(500, f"Error entrenando modelo: {str(e)}")
```

### 3. Router de predicción

**Archivo:** `zibooka-ml/app/routers/predict.py`

```python
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from app.models.demand_predictor import DemandPredictor
from app.config import settings
import os

router = APIRouter()

class DemandRequest(BaseModel):
    days_ahead: int = 30
    limit: Optional[int] = 20
    min_score: Optional[float] = 0.5

class DemandItem(BaseModel):
    product_id: str
    title: str
    category: str
    demand_score: float
    predicted_loans: float

class DemandResponse(BaseModel):
    predictions: list[DemandItem]
    model_version: str
    model_metrics: dict

@router.post("/demand", response_model=DemandResponse)
async def predict_demand(request: DemandRequest):
    """Predice qué libros tendrán alta demanda en los próximos días."""
    model_path = os.path.join(settings.models_dir, "demand_predictor.joblib")
    
    if not os.path.exists(model_path):
        raise HTTPException(404, "Modelo no encontrado. Ejecute /train/demand primero.")
    
    predictor = DemandPredictor.load(model_path)
    
    # Obtener datos actuales y predecir
    # ...
    
    return DemandResponse(
        predictions=[],
        model_version="1.0",
        model_metrics={}
    )
```

## Uso en Frontend (Admin)

Nuevo componente que muestra el panel de predicciones:

```jsx
// client/src/components/admin/DemandPredictions.jsx
const DemandPredictions = () => {
  const [predictions, setPredictions] = useState([]);
  
  useEffect(() => {
    axios.post('/api/prediction/demand', { days_ahead: 30 })
      .then(res => setPredictions(res.data.predictions))
      .catch(console.error);
  }, []);
  
  return (
    <div className="p-4">
      <h2 className="text-xl font-bold mb-4">Predicción de Demanda — Próximos 30 Días</h2>
      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
        {predictions.map(p => (
          <div key={p.product_id} className="border rounded p-3">
            <h3 className="font-semibold">{p.title}</h3>
            <p>Categoría: {p.category}</p>
            <div className="mt-2">
              <div className="bg-gray-200 rounded h-2">
                <div 
                  className="bg-blue-500 rounded h-2" 
                  style={{ width: `${p.demand_score * 100}%` }}
                />
              </div>
              <span className="text-sm">
                {Math.round(p.demand_score * 100)}% probabilidad
              </span>
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};
```

## Criterios de Aceptación

- [ ] `POST /train/demand` completa el entrenamiento con datos reales de MongoDB
- [ ] Retorna métricas: accuracy, precision, recall, F1, AUC-ROC
- [ ] Guarda el modelo serializado con joblib
- [ ] `POST /predict/demand` retorna predicciones con scores entre 0 y 1
- [ ] El modelo tiene F1 > 0.7 en datos de prueba (umbral mínimo aceptable)
- [ ] Las predicciones incluyen feature importance para interpretabilidad
- [ ] El endpoint tiene fallback si el modelo no está entrenado (error claro)
- [ ] Hay un panel admin en el frontend que muestra las predicciones
- [ ] Se almacena el log de métricas para tracking histórico

## Métricas de Referencia

| Métrica | Objetivo | Mínimo Aceptable |
|---------|----------|------------------|
| Accuracy | > 0.80 | 0.70 |
| Precision | > 0.75 | 0.65 |
| Recall | > 0.70 | 0.60 |
| F1 Score | > 0.72 | 0.62 |
| AUC-ROC | > 0.85 | 0.75 |
